openmls/src/treesync/treekem.rs

```diff
@@ -0,0 +1,314 @@
+            update_path_nodes.push(update_path_node);
+        }
+
+        Ok(UpdatePath {
+            leaf_key_package: key_package.clone(),
+            nodes: update_path_nodes.into(),
+        })
+    }
+
+    /// Decrypt an [`UpdatePath`] originating from the given
+    /// `sender_leaf_index`. The `group_context` is used in the decryption
+    /// process and the `exclusion_list` is used to determine the position of
+    /// the ciphertext in the `UpdatePath` that we can decrypt.
+    ///
+    /// Returns a vector containing the decrypted [`ParentNode`] instances, as
+    /// well as the [`CommitSecret`] resulting from their derivation. Returns an
+    /// error if the `sender_leaf_index` is outside of the tree.
+    pub(crate) fn decrypt_path(
+        &self,
+        backend: &impl OpenMlsCryptoProvider,
+        ciphersuite: &'static Ciphersuite,
+        version: ProtocolVersion,
+        update_path: &UpdatePath,
+        sender_leaf_index: LeafIndex,
+        exclusion_list: &HashSet<&LeafIndex>,
+        group_context: &[u8],
+    ) -> Result<(Vec<ParentNode>, CommitSecret), TreeKemError> {
+        let path_position = self.subtree_root_position(sender_leaf_index, self.own_leaf_index())?;
+        let update_path_node = update_path
+            .nodes()
+            .get(path_position)
+            .ok_or(TreeKemError::UpdatePathNodeNotFound)?;
+
+        let (decryption_key, resolution_position) =
+            self.decryption_key(sender_leaf_index, exclusion_list)?;
+        let ciphertext = update_path_node
+            .encrypted_path_secrets(resolution_position)
+            .ok_or(TreeKemError::EncryptedCiphertextNotFound)?;
+
+        let path_secret = PathSecret::decrypt(
+            backend,
+            ciphersuite,
+            version,
+            ciphertext,
+            decryption_key,
+            group_context,
+        )?;
+
+        let remaining_path_length = update_path.nodes().len() - path_position;
+        let (mut derived_path, _plain_update_path, commit_secret) =
+            ParentNode::derive_path(backend, ciphersuite, path_secret, remaining_path_length)?;
+        // We now check that the public keys in the update path and in the
+        // derived path match up.
+        for (update_parent_node, derived_parent_node) in update_path
+            .nodes()
+            .iter()
+            .skip(path_position)
+            .zip(derived_path.iter())
+        {
+            if update_parent_node.public_key() != derived_parent_node.public_key() {
+                return Err(TreeKemError::PathMismatch);
+            }
+        }
+
+        // Finally, we append the derived path to the part of the update path
+        // below the first node that we have a private key for.
+        let mut path: Vec<ParentNode> = update_path
+            .nodes()
+            .iter()
+            .take(path_position)
+            .map(|update_path_node| update_path_node.public_key().clone().into())
```

Comment:
Is it necessary that `update_path` is a reference? If not we could avoid cloning here.

Reply:
Good catch! I checked and it turns out we can avoid cloning here and in a few other places. I refactored accordingly.